### PR TITLE
Use Webpack to manage assets

### PR DESCRIPTION
**Pull request by tyom**
- Webpack loads Sass files
- Webpack loads fonts and images referenced in packages or `/assets/images` directory
- New `getAssetPath` function in global `locals` to get path to compiled asset
- Optimise images on build
- Refactor npm run scripts
- Integrate BrowserSync with Webpack
- Add asset revision hashes

Addresses #206

### CODE REVIEW

File: assets/stylesheets/application.scss [hunk 1]
```diff
@@ -1,7 +1,9 @@
 $images-dir: "/images/";
 
-@import "trade-elements";
-@import "font-awesome";
+@import "../../node_modules/@uktrade/trade_elements/dist/sass/trade-elements";
```
Comment (teneightfive) [inline assets/stylesheets/application.scss]: Can we not also do `~@uktrade/trade_elements/...`?

File: assets/stylesheets/objects/layout/_global-header.scss [hunk 2]
```diff
@@ -24,7 +24,7 @@
 }
 
 .global-header__crest {
-  background: url("../../assets/images/dit-crest__white.png") 6px 0 no-repeat $black;
+  background: url("../../../images/dit-crest__white.png") 6px 0 no-repeat $black;
```
Comment (teneightfive) [inline assets/stylesheets/objects/layout/_global-header.scss]: This nesting seems quite deep. Shouldn't we have stylesheets and images at the same level? 

Something like this:
```
/assets
  /stylesheets
    application.css
  /images
    ...
```

File: package.json [hunk 3]
```diff
@@ -1,11 +1,11 @@
 {
-  "name": "data-hub-frontend",
+  "name": "data-hub-fe-beta2",
```
Comment (teneightfive) [inline package.json]: Might want to rebase on top of develop as this reverts and earlier change from @feedmypixel

File: assets/stylesheets/objects/layout/_global-header.scss (hunk 2 above)
Comment (tyom) [inline assets/stylesheets/objects/layout/_global-header.scss]: They are like that, aren't they?

File: package.json (hunk 3 above)
Comment (tyom) [inline package.json]: Hmm.. It is up to date with `develop`.

File: src/server.js [hunk 4]
```diff
@@ -106,6 +106,7 @@ app.use(favicon(path.join(config.root, 'public/assets/images', 'favicon.ico')))
 app.use(express.static(path.join(config.root, 'public')))
 app.use('/javascripts', express.static(path.join(config.root, 'build/javascripts')))
 app.use('/css', express.static(path.join(config.root, 'build/css')))
+app.use('/images', express.static(path.join(config.root, 'build/images')))
```
Comment (teneightfive) [inline src/server.js]: Can we condense these 3 lines to be?

```
app.use('/assets', express.static(path.join(config.root, 'build')))
```

File: package.json [hunk 5]
```diff
@@ -96,7 +95,8 @@
     "css:autoprefixer": "postcss build/css/*.css --use autoprefixer -d build/css",
     "sync": "browser-sync start --proxy http://localhost:3000 --files build/css/*.css build/javascripts/*.js",
     "circle": "npm run lint && nyc --reporter=lcov npm run test:unit -- --reporter mocha-circleci-reporter && npm run sass && npm run js && codeclimate-test-reporter < coverage/lcov.info",
-    "heroku-postbuild": "npm rebuild node-sass && npm run sass && npm run js"
+    "heroku-postbuild": "npm rebuild node-sass && npm run sass && npm run js",
+    "watch:test": "npm run test:unit -- -w"
```
Comment (teneightfive) [inline package.json]: Do we want to use this watch pattern for `client` and `js`, etc?

File: package.json (hunk 3 above)
Comment (tyom) [inline package.json]: Somehow overwritten. I'll fix.

File: src/views/_layouts/dit-base.njk [hunk 6]
```diff
@@ -1,4 +1,4 @@
-{%- set assetPath = asset_path | default('/') -%}
+{%- set publicPath = asset_path | default('/') -%}
```
Comment (teneightfive) [inline src/views/_layouts/dit-base.njk]: Should we try and remove `asset_path` and `publicPath` all together and have 1 way to load assets?

Maybe a function on locals like `assetPath('filename.ext')` which checks the manifest rather than a filter?

File: assets/stylesheets/objects/layout/_global-header.scss (hunk 2 above)
Comment (teneightfive) [inline assets/stylesheets/objects/layout/_global-header.scss]: Not sure why we need to go back 3 directories then?

File: src/views/_layouts/dit-base.njk (hunk 6 above)
Comment (tyom) [inline src/views/_layouts/dit-base.njk]: `assetPath` is not the same as `publicPath`. One is Webpack the other is Express.

File: package.json (hunk 3 above)
Comment (teneightfive) [inline package.json]: Maybe bin these changes. You probably amended it wrong during a conflict.

File: src/views/_layouts/dit-base.njk (hunk 6 above)
Comment (tyom) [inline src/views/_layouts/dit-base.njk]: Started with a function, then switched to filter. No I look at it and think that maybe function call is better.

File: assets/stylesheets/objects/layout/_global-header.scss (hunk 2 above)
Comment (tyom) [inline assets/stylesheets/objects/layout/_global-header.scss]: Because `_global-header` is in `stylesheets/objects/layout/`.

File: src/views/_layouts/dit-base.njk (hunk 6 above)
Comment (teneightfive) [inline src/views/_layouts/dit-base.njk]: I think we should load assets the same way though. In `datahub-base.njk` we're doing:

```njk
<script src="{{ 'ie.js' | asset }}"></script>
```

but in `dit-base.njk` we're doing:

```njk
<link href="{{ publicPath }}css/application.css" media="screen" rel="stylesheet">
```

Comment (teneightfive) [inline src/views/_layouts/dit-base.njk]: I think there are also some references to js files in this file that may have been missed

File: assets/stylesheets/objects/layout/_global-header.scss (hunk 2 above)
Comment (teneightfive) [inline assets/stylesheets/objects/layout/_global-header.scss]: So this isn't looking for the compiled path?

Comment (tyom) [inline assets/stylesheets/objects/layout/_global-header.scss]: No, this is taking source asset and Webpack then substitutes it with the compiled one.

**Comment by tyom** [comment]
> Following on from what Ben said. I think that this is best done after we merge trade elements in 

Sure.

> and I'm not convinced there is a benefit to handling styles in webpack right now and the whole setting the whole app styles as a dependency in app.js feels off.
> Why don't we discuss styles in general after we;ve reorganised the code and are starting on components as it would be way nicer to use weback and have each component define it's style dependency.

The benefit of doing it in Webpack is to have single place to configure and manage asset dependencies. Hash them, configure them.

Going forward we may decide to reference styles from within individual components.

File: src/server.js (hunk 4 above)
Comment (tyom) [inline src/server.js]: I removed `/assets` from the URLs. Felt it was unnecessary.

File: config/nunjucks/globals.js [hunk 7]
```diff
@@ -1,6 +1,27 @@
...
+}
+
 module.exports = {
   serviceTitle: 'Data Hub',
   projectPhase: 'alpha',
   description: 'Data Hub is a customer relationship, project management and analytical tool for Department for International Trade.',
   feedbackLink: '/support/bug',
+
+  assetPath (asset) {
+    const assetsUrl = process.env.assetsUrl || this.ctx.BASE_URL || '/'
```
Comment (teneightfive) [inline config/nunjucks/globals.js]: We should try and keep properties set on `process.env` in one place in the config file so its easier to see what env vars can be set. 

We should also use const case for environment variables.

File: config/nunjucks/globals.js [hunk 8]
```diff
@@ -1,6 +1,27 @@
...
+  logger.error('Manifest file is not found. Ensure assets are built.')
+}
+
 module.exports = {
   serviceTitle: 'Data Hub',
   projectPhase: 'alpha',
   description: 'Data Hub is a customer relationship, project management and analytical tool for Department for International Trade.',
   feedbackLink: '/support/bug',
+
+  assetPath (asset) {
```
Comment (teneightfive) [inline config/nunjucks/globals.js]: I'm not sure `globals.js` is the best place for this method. Wondering whether it should be its own middleware or set in the locals middleware?

File: config/nunjucks/globals.js (hunk 7 above)
Comment (teneightfive) [inline config/nunjucks/globals.js]: If we're adding new env vars we should explain them in the readme too.

File: src/server.js [hunk 9]
```diff
@@ -28,6 +25,8 @@ const logger = require('../config/logger')
 
 const router = require('../config/routes')
 
+const { redisStore } = require('./store')
```
Comment (teneightfive) [inline src/server.js]: Move to config, like the logger and where we might end up having express config?

Comment (teneightfive) [inline src/server.js]: Making call it something more explicit like `session-store.js` too.

Comment (teneightfive) [inline src/server.js]: Feels like this would also be neater as a separate PR. This PR is snaking quite a bit now...

Comment (tyom) [inline src/server.js]: Moved out in #262.

File: config/nunjucks/globals.js (hunk 8 above)
Comment (tyom) [inline config/nunjucks/globals.js]: We could make a middleware to automatically match asset paths to those defined in manifest and return them if found but that means that all assets would need to be referenced by their chunk name, e.g. `app.js` instead of full path. And I feel that it might be a bit too magical. It might be better to grasp what's going on by having an explicit global Nunjucks function. At least for now.

Comment (teneightfive) [inline config/nunjucks/globals.js]: I didn't mean make a magic middleware I just meant I don't think it should live as a nunjucks _global_. 

Feels better suited in `locals` middleware of a separate middleware that sets a method on `res.locals`?

File: config/nunjucks/globals.js [hunk 10]
```diff
@@ -1,6 +1,28 @@
...
+    const assetsUrl = config.assetsHost || this.ctx.BASE_URL || '/'
```
Comment (teneightfive) [inline config/nunjucks/globals.js]: Also because this is falling back on the `BASE_URL` value which is set in locals anyway. Maybe it would be better in the same location.

File: webpack.config.js [hunk 11]
```diff
@@ -1,24 +1,78 @@
 const webpack = require('webpack')
+const merge = require('webpack-merge')
 const path = require('path')
+const ExtractTextPlugin = require('extract-text-webpack-plugin')
+const WebpackAssetsManifest = require('webpack-assets-manifest')
+const BrowserSyncPlugin = require('browser-sync-webpack-plugin')
 
 const isProd = process.env.NODE_ENV === 'production'
```
Comment (teneightfive) [inline webpack.config.js]: Use `!config.isDev`? Trying to reduce using `process.env` all over the app.

Comment (teneightfive) [inline webpack.config.js]: Use `!config.isDev`? Trying to reduce using `process.env` all over the app.

File: src/server.js [hunk 12]
```diff
@@ -53,14 +53,12 @@ app.set('view engine', 'njk')
...
-app.use(favicon(path.join(config.root, 'public/assets/images', 'favicon.ico')))
+app.use(favicon(path.join(config.root, 'public/images', 'favicon.ico')))
 app.use(express.static(path.join(config.root, 'public')))
 app.use('/javascripts', express.static(path.join(config.root, 'build/javascripts')))
 app.use('/css', express.static(path.join(config.root, 'build/css')))
-
-app.use('/images', express.static(`${__dirname}/../assets/images/_deprecated/`))
-
-app.use('/fonts', express.static(`${__dirname}/../node_modules/font-awesome/fonts`))
+app.use('/images', express.static(path.join(config.root, 'build/images')))
```
Comment (teneightfive) [inline src/server.js]: Wondering whether `build` should be a dot folder (`.build`)? Might be easier to see it contains generated assets and keeps the output folder at the top of the list in most IDEs.

File: webpack.config.js [hunk 13]
```diff
@@ -27,17 +81,38 @@ const webpackConfig = {
...
+      ],
+    }, {
+      reload: false,
+    }),
+  ],
+})
 
-  webpackConfig.plugins.push(
+const prod = merge.smart(common, {
+  devtool: 'hidden-source-map',
```
Comment (teneightfive) [inline webpack.config.js]: Seems to still be outputting sourcemap for CSS in production build mode

File: package.json (hunk 5 above)
Comment (tyom) [inline package.json]: Not sure what you mean.

File: config/index.js [hunk 14]
```diff
@@ -40,3 +41,7 @@ module.exports = {
   longDateFormat: 'D MMMM YYYY',
   mediumDateFormat: 'D MMM YYYY',
 }
+
+config.buildDir = path.join(config.root, '.build')
```
Comment (teneightfive) [inline config/index.js]: What about moving the root out to a var so this can still defined in the object?

```js
const root = path.normalize(`${__dirname}/..`)

const config = {
  root,
  ...
  buildDir: path.join(root, '.build'),
}
```

File: nodemon.json [hunk 15]
```diff
@@ -0,0 +1,13 @@
+{
+  "restartable": "rs",
+  "watch": [
+    "src/**/*.js"
+  ],
+  "ignore": [
+    "build/"
```
Comment (teneightfive) [inline nodemon.json]: `".build/"`?

File: src/server.js [hunk 16]
```diff
@@ -53,14 +53,12 @@ app.set('view engine', 'njk')
...
-app.use('/javascripts', express.static(path.join(config.root, 'build/javascripts')))
-app.use('/css', express.static(path.join(config.root, 'build/css')))
-
-app.use('/images', express.static(`${__dirname}/../assets/images/_deprecated/`))
-
-app.use('/fonts', express.static(`${__dirname}/../node_modules/font-awesome/fonts`))
+app.use('/js', express.static(path.join(config.buildDir, 'js')))
+app.use('/css', express.static(path.join(config.buildDir, 'css')))
+app.use('/images', express.static(path.join(config.buildDir, 'images')))
+app.use('/fonts', express.static(path.join(config.buildDir, 'node_modules/font-awesome/fonts')))
```
Comment (teneightfive) [inline src/server.js]: What are we even still using fontawesome for?!

Might be another PR to see if we can remove the need for it...

File: webpack.config.js [hunk 17]
```diff
@@ -53,7 +130,12 @@ if (isProd) {
...
       dead_code: true,
     }),
-  )
-}
+    new ExtractTextPlugin('css/[name].[chunkhash:8].css'),
+  ],
+})
 
-module.exports = webpackConfig
+if (config.isDev) {
```
Comment (teneightfive) [inline webpack.config.js]: Wondering if your original way around was better? 

Should the default be prod or develop? To get the develop config this setup would rely on `NODE_ENV=development` being specifically set.

File: config/index.js (hunk 14 above)
Comment (tyom) [inline config/index.js]: Good idea. Much better!

File: src/server.js (hunk 16 above)
Comment (tyom) [inline src/server.js]: I agree. It's used for dropdown icons and maybe something else. We should get rid of it. Perhaps move to SVG icons.

File: webpack.config.js (hunk 17 above)
Comment (tyom) [inline webpack.config.js]: I've another condig prop with explicit checking of prod.

File: config/index.js [hunk 18]
```diff
@@ -1,6 +1,7 @@
 const path = require('path')
 
 const isDev = (process.env.NODE_ENV !== 'production')
+const isProd = (process.env.NODE_ENV === 'production')
```
Comment (teneightfive) [inline config/index.js]: This can probably just be defined inline in the property definition.

The `isDev` variable is used elsewhere which is why it's defined outside the object.

File: src/middleware/locals.js [hunk 19]
```diff
@@ -4,20 +4,40 @@ const config = require('../../config')
 
 const startTime = new Date().getTime()
 
+let webpackManifest = {}
```
Comment (feedmypixel) [inline src/middleware/locals.js]: does this need a value assigned to it?

File: src/middleware/locals.js [hunk 20]
```diff
@@ -4,20 +4,40 @@ const config = require('../../config')
...
+try {
+  webpackManifest = require(`${config.buildDir}/manifest.json`)
+} catch (err) {
+  logger.error('Manifest file is not found. Ensure assets are built.')
```
Comment (feedmypixel) [inline src/middleware/locals.js]: Is having no manifest a valid option? does this need to return?

File: src/middleware/locals.js [hunk 21]
```diff
@@ -4,20 +4,40 @@ const config = require('../../config')
...
+    startTime: startTime,
+    referer: req.headers.referer,
+    env: config.env,
+    googleTagManagerKey: config.googleTagManagerKey,
+    query: req.query,
+    currentPath: req.path,
+    BASE_URL: baseUrl,
+    CANONICAL_URL: baseUrl + req.originalUrl,
+
+    getAssetPath (asset) {
```
Comment (feedmypixel) [inline src/middleware/locals.js]: going forward maybe this could be moved into its own file in a `src/middleware/locals` folder?

File: webpack.config.js [hunk 22]
```diff
@@ -1,24 +1,78 @@
...
+      },
+      {
+        test: /\.scss$/,
+        use: ExtractTextPlugin.extract({
+          fallback: 'style-loader',
+          use: [
+            {
+              loader: 'css-loader',
+              options: {
+                sourceMap: config.isDev,
```
Comment (feedmypixel) [inline webpack.config.js]: do we need both of these in config? Can we just negate `config.isProd`? potentially for a different PR

File: webpack.config.js [hunk 23]
```diff
@@ -27,17 +81,38 @@ const webpackConfig = {
...
-if (isProd) {
-  webpackConfig.devtool = 'hidden-source-map'
+const develop = merge.smart(common, {
+  output: {
+    filename: 'js/[name].js',
+  },
+  plugins: [
+    new ExtractTextPlugin('css/[name].css'),
+    new BrowserSyncPlugin({
+      port: 3001,
```
Comment (feedmypixel) [inline webpack.config.js]: should this be referenced from config?

File: webpack.config.js [hunk 24]
```diff
@@ -27,17 +81,38 @@ const webpackConfig = {
...
+      proxy: 'http://localhost:3000',
```
Comment (feedmypixel) [inline webpack.config.js]: should this be referenced from config?

File: webpack.config.js [hunk 25]
```diff
@@ -50,10 +125,15 @@ if (isProd) {
...
     }),
-  )
-}
+    new ExtractTextPlugin('css/[name].[chunkhash:8].css'),
+  ],
+})
 
-module.exports = webpackConfig
+if (config.isProd) {
+  module.exports = prod
```
Comment (feedmypixel) [inline webpack.config.js]: sat here by themselves `prod` and `develop` feel like they need a little more explanation. Maybe `prodConfig` and `devConfig`?

File: config/index.js [hunk 26]
```diff
@@ -30,6 +33,7 @@ module.exports = {
     // 2 hour timeout
     ttl: process.env.SESSION_TTL || (2 * 60 * 60 * 1000),
   },
+  assetsHost: process.env.ASSETS_HOST,
```
Comment (tyom) [inline config/index.js]: I'm getting the base URL in locals middleware where there's special logic to obtain it.

File: src/middleware/locals.js (hunk 19 above)
Comment (tyom) [inline src/middleware/locals.js]: Makes it easier to deal with missing property when we do `webpackManifest[asset]` in `getAssetPath`. Just returns `undefined` instead of throwing an error.

File: src/middleware/locals.js (hunk 20 above)
Comment (tyom) [inline src/middleware/locals.js]: It is a valid option if you wish to serve static files from other Express static locations (i.e. `public/`)

File: webpack.config.js (hunk 22 above)
Comment (tyom) [inline webpack.config.js]: I find it easier to read and the logic for them is quite different.

File: webpack.config.js (hunk 23 above)
Comment (tyom) [inline webpack.config.js]: Nah. This is BrowserSync spefic and is only used in this one place.

File: webpack.config.js (hunk 25 above)
Comment (tyom) [inline webpack.config.js]: Hmm.. You are in a file called `webpack.config.js`. Feel like that's already setting the scope for what this is.
